### PR TITLE
Broadcast game state through web sockets

### PR DESCRIPTION
**Pull request by LeNitrous**
Addresses https://github.com/ppy/osu/discussions/13011 and supersedes https://github.com/ppy/osu/pull/12274.

- This adds a new setting under "Integration Settings" to enable or disable game state broadcasting.
- When enabled, the game state broadcaster can be accessed through `ws://localhost:7270/state/`.
- The broadcaster currently emits ruleset, beatmap, user activity changes as well as live statistics during gameplay.

Example:

https://user-images.githubusercontent.com/20495991/167238619-f579a08e-86f5-4193-9d46-bc662a225f74.mp4

Adds `WebSocketServer`, `GameStateBroadcastServer` and `GameStateBroadcaster` components.
- `WebSocketServer` is an abstract `CompositeDrawable` capable of listening and accepting web socket requests.
- `GameStateBroadcastServer` hosts child `GameStateBroadcaster` components which allows dependency injection and allows removal and adding of children as needed.
- `GameStateBroadcaster` is an abstract `Component` capable of broadcasting JSON text of the current game's state. The JSON payload broadcasted uses this schema:

```json
{
    "Type": "<string>",
    "Message": "<any>"
}
```

Differences with the original PR:
- This reliably handles disconnections requested by the server or client.
- The broadcasters have been modularized for extensibility and ease of integration instead of it being a large JSON payload.

Example client in Javascript:
```js
var socket = new WebSocket("ws://localhost:7270/state/");
socket.onmessage = function(e) {
    let { Type, Message } = JSON.parse(e.data);
    console.log(Type);      // string
    console.log(Message);   // any
}
```

### CODE REVIEW

**Comment by LeNitrous** [comment]
Previously discussed on Discord, `HttpListener` is considered as deprecated API and only exists for compatibility reasons. Microsoft encourages use of ASP.NET. However upon testing in an isolated project, attempting to run a Kestrel server with minimal APIs using `Microsoft.AspNetCore.Server.Kestrel 2.2.0` will fail to run if `Microsoft.AspNetCore.SignalR.Client` is included as a package reference. Should we still be using `HttpListener` in the meantime?

**Comment by frenzibyte** [comment]
Sounds like you might want to report this [against their side](https://github.com/dotnet/aspnetcore).

**Comment by LeNitrous** [comment]
Reported the issue (https://github.com/dotnet/aspnetcore/issues/41575) with the ASP.NET team and according to the respondents, it is a valid issue but its also a "wont fix" case as the Kestrel package on NuGet is deprecated. This is rather unfortunate as the only real way left is implementing it from scratch using a `TcpListener` which at this point goes beyond the scope of this PR.

**Comment by davidfowl** [comment]
I don't know what the constraints are but why not target .NET 6?

**Comment by peppy** [comment]
> I don't know what the constraints are but why not target .NET 6?

This is our eventual goal, but we are currently limited by mobile platforms (where we still rely on xamarin). That said, you make a valid point in this case as the code being changed here likely only needs to work on desktop platforms, where we can conditionally target .NET 6.

I haven't gone through this PR yet, but @LeNitrous you may be able to move the implementation to the `osu.Desktop` project and isolate to .NET 6, if that helps.

**Comment by smoogipoo** [comment]
It's not that we can't use the full ASP.NET framework, but the contributor above was somewhat concerned with keeping the footprint of our application as low as possible by not pulling in too many unnecessary dependencies.

I've tested a simple dummy application and a publish inclusive of ASP.NET adds 20MB to the output. I'm not sure if this is unreasonable just yet, and it's also an isolated test so potentially many dependencies would duplicate existing ones, but it's something to keep in mind.

Personally, I'm not super against pulling in ASP.NET if there's no lightweight alternative to websockets fit for our needs, and waiting on assembly trimming/linking for the future.

**Comment by davidfowl** [comment]
No problem. If you care about the performance, then you should use ASP.NET Core but I don't know what the constraints are so I'll leave you to make the best decisions for your game 😄

**Comment by LeNitrous** [comment]
I have taken up @peppy's suggestion in moving the the websocket server to `osu.Desktop` and use the `Microsoft.AspNetCore.App` framework to use Kestrel. Additionally, I removed the ability for the game state broadcaster to broadcast player state changes until I find an elegant solution (or may be included in a separate PR instead).

**Comment by davidfowl** [comment]
Do you want any feedback on the websocket implementation? How much do you care about performance?

**Comment by bastianpedersen** [comment]
Is this a feature that will be implemented in the future? Seems like the discussion is quite dead.

**Comment by frenzibyte** [comment]
There are no discussions left in this PR, this is only pending review.

**Comment by LeNitrous** [comment]
Preliminary analysis on possible solutions based on my findings and suggestions made in Discord on how to deal with ASP.NET.
### Current Situation
Currently, to be able to use a Websocket Server, the Kestrel Server is needed to be installed as a package dependency. However, installing the dependency alone is not enough as the entirety of the ASP.NET Core framework is required. This introduces unnecessary bloat to osu! itself as we are only interested in a Websocket broadcast server.

#### Solution 1: Enable Trimming
This may affect custom rulesets as enabling trimming [disallows dynamic assembly loading](https://learn.microsoft.com/en-us/dotnet/core/deploying/trimming/incompatibilities#dynamic-assembly-loading-and-execution).
#### Solution 2: Write a websocket server from scratch
This may be an overkill solution and I highly doubt we may want to maintain this.
#### Solution 3: Make use of a third-party implementation
[Watson Websocket](https://github.com/jchristn/WatsonWebsocket) is an MIT-Licensed third-party implementation of the Websocket standard which is around 200 KB in package size (as stated in the NuGet page) and does not need any more additional dependencies. This may be one of the feasible solutions I can suggest.

Although I do want to hear thoughts from others about the solutions proposed before I proceed in making changes.

**Comment by peppy** [comment]
> This may affect custom rulesets as enabling trimming [disallows dynamic assembly loading](https://learn.microsoft.com/en-us/dotnet/core/deploying/trimming/incompatibilities#dynamic-assembly-loading-and-execution).

One would hope with the template and built-in rulesets we would have 100% coverage of ruleset API code. If not, we would probably need to forcefully exclude it from trimming (I'd hope this is possible). This does seem like the best path forward though.

**Comment by SirWaddles** [comment]
I thought I'd give this a try since I stream osu sometimes and was previously using gosumemory for this. I did make a merge to master, but I'm not very experienced with .NET 6 so I probably made a few mistakes.

It works really well! It's nice that this could be a feature native to osu. The player state I think is missing at the moment though.

I'm not sure if this can be accommodated, but in the past I had my websocket bound to my local address rather than localhost. This is because I'm using a separate streaming PC and I had it running the overlay. I could set up a proxy if need be, but I thought I'd mention it.

Thanks for building this @LeNitrous

**Comment by ilsubyeega** [comment]
I would like to present a different perspective. 
Currently, I understand that this feature is implemented by building a WebSocket server to broadcast the data. This feature is limited to the osu.Desktop platform. 
If it is tricky to create a WebSocket server, an alternative approach worth considering could be implementing this feature as a client instead of a server. Of course, this would require a separate third-party WebSocket server, but I believe this would allow this feature to be fully utilized not only on the osu.Desktop platform, but also on mobile devices such as smartphones and tablets.

**Comment by longnguyen2004** [comment]
Are you still working on this? I think using a third-party library is the way to go, but we should ask for dev team's opinion

**Comment by smoogipoo** [comment]
This is currently still low priority for the time being.

**Review by Krown9 — APPROVED** [review]

**Comment by Akarinnnnn** [comment]
> Previously discussed on Discord, `HttpListener` is considered as deprecated API and only exists for compatibility reasons. Microsoft encourages use of ASP.NET. However upon testing in an isolated project, attempting to run a Kestrel server with minimal APIs using `Microsoft.AspNetCore.Server.Kestrel 2.2.0` will fail to run if `Microsoft.AspNetCore.SignalR.Client` is included as a package reference. Should we still be using `HttpListener` in the meantime?

We can use `WebSocket.CreateFromStream` after .NET 6. Such implementation is provided by assembly `System.Net.WebSockets` and doesn't requires Kestrel. In case we can't add ASP.NET Core as a dependency, it can be used as alternative.

**Comment by Chiffario** [comment]
> Previously discussed on Discord, `HttpListener` is considered as deprecated API and only exists for compatibility reasons. Microsoft encourages use of ASP.NET. However upon testing in an isolated project, attempting to run a Kestrel server with minimal APIs using `Microsoft.AspNetCore.Server.Kestrel 2.2.0` will fail to run if `Microsoft.AspNetCore.SignalR.Client` is included as a package reference. Should we still be using `HttpListener` in the meantime?

This got a bit lost in out-of-Github discussions, but, as was discovered, [HttpListener isn't deprecated but just frozen](https://github.com/dotnet/runtime/issues/63941#issuecomment-1205259894). After mentioning this to @bdach back in October, he said that using `HttpListener` shouldn't be an issue, as newer APIs and features aren't necessary for this use-case

**Comment by LeNitrous** [comment]
Upon further investigation, `HttpListener` is feasible however with one caveat when ran under Windows. Due to its dependence on `http.sys`, it requires ACLs to be set up or else it would just throw "Access Denied" exceptions. This in turn requires another workaround where either the user or we call `netsh` (or use `httpapi.dll` in code) to allow `HttpListener` to listen to specified URLs but either method requires elevated permissions.

**Comment by Akarinnnnn** [comment]
I tested on a win11, the add prefix step didn't throw exception.



---Original---
From: ***@***.***&gt;
Date: Sat, May 10, 2025 18:52 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [ppy/osu] Broadcast game state through web sockets (PR #18129)


 LeNitrous left a comment (ppy/osu#18129)
 
Upon further investigation, HttpListener is feasible however with one caveat when ran under Windows. Due to its dependence on http.sys, it requires ACLs to be set up or else it would just throw "Access Denied" exceptions. This in turn requires another workaround where either the user or we call netsh (or use httpapi.dll in code) to allow HttpListener to listen to specified URLs but either method requires elevated permissions.

File: osu.Game/Configuration/OsuConfigManager.cs [hunk 1]
```diff
@@ -227,6 +227,8 @@ protected override void InitialiseDefaults()
             SetDefault(OsuSetting.EditorSubmissionLoadInBrowserAfterSubmission, true);
 
             SetDefault(OsuSetting.WasSupporter, false);
+            SetDefault(OsuSetting.EditorWaveformOpacity, 0.25f);
```
Comment (smoogipoo) [inline osu.Game/Configuration/OsuConfigManager.cs:230]: Seems newly added in this PR.

File: osu.Game/Online/Broadcasts/Broadcaster.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+        protected void Broadcast(string type, object message)
+        {
+            var payload = new Payload
+            {
+                Type = type,
+                Message = message,
+            };
+
+            server?.Broadcast(payload.Serialize()).WaitSafely();
```
Comment (smoogipoo) [inline osu.Game/Online/Broadcasts/Broadcaster.cs:30]: Should not block.

File: osu.Game/Online/Broadcasts/GameplayStateBroadcaster.cs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+                Name = mod.Acronym,
+                Settings = mod.GetOrderedSettingsSourceProperties().Select(pair => new ModSettingDetails
+                {
+                    Name = pair.Item1.Label.ToString(),
+                    Value = pair.Item2.GetValue(mod),
+                }).ToArray()
+            }).ToArray());
+
+            broadcast(@"ruleset", state.Ruleset.RulesetInfo.ShortName);
+            broadcast(@"beatmap", state.Beatmap.BeatmapInfo);
```
[the remark this answers is not in the document]
Comment (LeNitrous) [reply osu.Game/Online/Broadcasts/GameplayStateBroadcaster.cs:39]: I would probably have to split this to two parts wherein this PR only focuses implementing the websocket client and server and actually broadcasting data in another seperate PR. I agree it has to conform to an agreed upon structure while also implementing things like debouncing. Splitting the PR may also introduce the possibility of discussion what data needs to be sent and how frequent it should be sent.

**Comment by Akarinnnnn** [comment]
For debounce, I think we can poll these broadcasters every 150ms, do by WebsocketServer. Another performance impact point is, now we serialize data into UTF-8 internally first, then convert buffer into UTF-16 string, finally convert them back UTF-8 again. Obviously it costs.



---Original---
From: "Dan ***@***.***&gt;
Date: Sun, May 11, 2025 19:16 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [ppy/osu] Broadcast game state through web sockets (PR #18129)




 
@smoogipoo commented on this pull request.
 
 
In osu.Game/Online/Broadcasts/GameplayStateBroadcaster.cs:
 &gt; +            broadcast(@"mods", state.Mods.Select(mod =&gt; new ModDetails +            { +                Name = mod.Acronym, +                Settings = mod.GetOrderedSettingsSourceProperties().Select(pair =&gt; new ModSettingDetails +                { +                    Name = pair.Item1.Label.ToString(), +                    Value = pair.Item2.GetValue(mod), +                }).ToArray() +            }).ToArray());  
Can we broadcast plain APIMod instead?
 
 
In osu.Game/Online/Broadcasts/GameplayStateBroadcaster.cs:
 &gt; +            broadcast(@"ruleset", state.Ruleset.RulesetInfo.ShortName); +            broadcast(@"beatmap", state.Beatmap.BeatmapInfo);  
This is weird... Can we broadcast objects that look like this instead?
 "gameplay_settings": {     "beatmap" : ...,     "ruleset": ...,     "mods": [...] } "gameplay_state": {     "score": ...,     "health": ... }  
Right now I'm not even sure how we're supposed to document this. At least having a clearly defined structure will help people piece together how to actually use it.
 
Also, this brings up the concern of performance. I originally had gameplay_state inside the former object before realising that would reserialise everything every time. Not only do we not want that, we also don't want to serialise things like health which update every frame (at least, while this is JSON), so this has to be debounced.
 
Performance needs to be a front-and-center concern with this entire thing. To the point that serialising as strings may not be good enough. We may want to consider MessagePack or a custom implementation based on Utf8JsonWriter, both with pooled byte arrays.
 
 
In osu.Game/Configuration/OsuConfigManager.cs:
 &gt; @@ -227,6 +227,8 @@ protected override void InitialiseDefaults()              SetDefault(OsuSetting.EditorSubmissionLoadInBrowserAfterSubmission, true);                SetDefault(OsuSetting.WasSupporter, false); +            SetDefault(OsuSetting.EditorWaveformOpacity, 0.25f);  
Seems newly added in this PR.
 
 
In osu.Game/Online/Broadcasts/Broadcaster.cs:
 &gt; + +        [BackgroundDependencyLoader] +        private void load(IBroadcastServer server) +        { +            this.server = server; +        } + +        protected void Broadcast(string type, object message) +        { +            var payload = new Payload +            { +                Type = type, +                Message = message, +            }; + +            server?.Broadcast(payload.Serialize()).WaitSafely();  
Should not block.

**Review by Stedoss — COMMENTED** [review]
A few things concern me here:

* Documentation is going to be a bit of a hassle, and will need continued maintenance as things change.
* As it stands right now, this sends an insane number of messages (imo) during gameplay - enough to almost crash HTTP testing apps like Insomnia or Hoppscotch. As others mentioned, needs to find a good balance of debounce there.
* As a potential consumer, this breaks a few things I would expect to happen. For example, going through the carousel and changing maps does not send any messages - it only happens on gameplay state change/user activity change (where you get a lot of `None`s). A consumer also does not receive any information on connection start, such as current activity - which the consumer would only know about when the activity changes _after_ connection start, whenever that would be. It just needs to be consistent and well-defined.
* Might be a little OTT, but I'd expect some sort of visual test so we are able to see actual messages being sent without going through gameplay.
* Broadcasts don't seem to be using well defined contracts - for example, the beatmap message just slams `BeatmapInfo` at the consumer with stuff it probably doesn't need.
* I feel there needs to be more considerations on _what_ actions send events. For example, editing a beatmap will send an `activity` message with content `ModdingBeatmap` (not really humanised either), but it does not send any content about the beatmap being edited, same when changing difficulties. Same when navigating multiplayer, I imagine people would want information about the room.

Just some quick thoughts from me, I'll keep an eye on this one.

File: osu.Game/Online/Broadcasts/GameplayStateBroadcaster.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (smoogipoo) [reply osu.Game/Online/Broadcasts/GameplayStateBroadcaster.cs:39]: I don't mind this being done right here, there's only two broadcasters and of them this is the only "complex" implementation.

Needs a bit more discussion which may result in core changes, but that's good to figure out right here imo :)

**Comment by LeNitrous** [comment]
I have removed broadcasting capabilities in this PR for now and focus on getting websocket communication functionality first.
